### ISSUE TITLE
Develop options to limit created Bitly short URLs.

### ISSUE THREAD

**Issue by nutsandbolts** [warfare-plugins/social-warfare#534]
# **Notes for Testers** 

We had been receiving complaints about Social Warfare eating up all 500 links on Bitly's free plan. 
The new Bitly options allow you to set a specific start date and select which post types will generate links. 

### What we want to happen: 
Social Warfare only creates Bitly links for URLs based on the admin's preferred settings. 

### Step 1: Register with Bitly

I recommend creating a test account for Bitly, so you do not use your own Bitly links for the month. 

Social Warfare -> Advanced -> Bitly Link Shortening:

1. Set "Bitly Link Shortening" to "ON."
2. Press "Save Changes." 
3. Click the "Authenticate" button and follow Bitlly's directions to register your account. 
4. Once you have finished Bitly's instructions, return to Social Warfare and reload the page.

### Step 2: Set up your Bitly Preferences

In Social Warfare -> Advanced -> Bitly Link Shortening:

1. In , set the start date to some time in the future. 
2. Select which post types should and should not create links. Make sure "post" is "ON."
3. Press "Save Changes."

### Step 3: Create a new post

We need to test on a post which we know has not yet been cached by Social Warfare. 

1. Create a new post with any kind of content. 
2. Visit the post on the front end. 
2. View your Bitly account. There should not be any links created for the post. 

### Step 4: Set the date back

In Social Warfare -> Advanced -> Bitly Link Shortening:

1. Set the date to some time in the past. 
2. Press "Save Changes."
3. Create another new post. 
4. View the post on the front end. 
5. You should see Bitly links generated for this post. 

### Step 5: Test on a Custom Post Type

If your theme or plugins do not have any Custom Post Types, such as a product page, you can skip this step. 

We want to test that Social Warfare will not generate links when "OFF" for this type, and will when after the start date and "ON" for the type. 

1. In Social Warfare -> Advanced -> Bitly Link Shortening, set the targeted post type to "OFF."
2. Press "Save Changes."
3. Create a new post of that type. 
4. View it on the front end. 
5. Visit Bitly. There should be no links generated for that post type. 

6. In Social Warfare -> Advanced -> Bitly Link Shortening, set the targeted post type to "ON."
7. Press "Save Changes."
8. Create a new post of that type. 
9. View it in the front end. 
10. Visit Bitly. There should be links generated for that post type. 

# 

If all of the above goes through, hooray, we are good to go! Give this a thumbs up. 
Else if the above does not work, please reply to this thread explaining what happened. 

If something is not clear or the process does not make sense, please use Slack to ask questions. 

**// end notes**
### Original issue

#

Currently when someone authenticates with Bitly, SW creates thousands and thousands of links for every page, post, etc. on the site. If the site is using a custom short domain for Bitly, they're limited to 500 custom shortlinks per month, which obviously SW eats up in a matter of seconds, and then they're stuck using the generic bit.ly links instead until the next month.

A couple of potential approaches here:

(1) We only create Bitly links for everything published from that point forward
(2) We allow the user to set a past date and create Bitly links for posts after that date only
(3) We refine the functionality so that Bitly links are only generated for social networks that allow them, like Twitter and Facebook (or maybe just Twitter since they really have no point on Facebook)
(4) We only create 100 or so Bitly links (starting with the most recent posts) to avoid hitting the person's limit
(5) We allow users to manually edit a page/post and click a "generate Bitly links" button, and nothing gets generated unless that happens

In other words, there are better ways to handle this than what we're doing right now. We need to discuss this further and decide how best to move forward, because what we're doing now causes a lot of complaints and confusion.

## Solution:
1. Create a text field that allows people to input a date as a string (don't worry about a date picker).
2. Create a dynamic list of post types where people can select on/off toggles for link sortening for each post type.
3. Only use link shortening on posts after that date unless the field is blank then do all, and only use them on post types where they are turned on.

**Comment by ckmahoney** [comment]
Of all of the options above, I think only implementing number 5 would offer the most control for the admin.

**Comment by ckmahoney** [comment]
This issue has been annotated and is ready for testing.
